src/Storages/ObjectStorage/DataLakes/Iceberg/IcebergWrites.cpp

```diff
@@ -0,0 +1,772 @@
+    {
+        if (!data_filenames.contains(partition_key))
+        {
+            auto data_filename = filename_generator.generateDataFileName();
+            data_filenames[partition_key] = data_filename;
+
+            const auto & settings = context->getSettingsRef();
+            const auto chosen_compression_method = chooseCompressionMethod(data_filename, configuration->compression_method);
+
+            auto buffer = object_storage->writeObject(
+                StoredObject(data_filename), WriteMode::Rewrite, std::nullopt, DBMS_DEFAULT_BUFFER_SIZE, context->getWriteSettings());
+
+            write_buffers[partition_key] = wrapWriteBufferWithCompressionMethod(
+                std::move(buffer),
+                chosen_compression_method,
+                static_cast<int>(settings[Setting::output_format_compression_level]),
+                static_cast<int>(settings[Setting::output_format_compression_zstd_window_log]));
+
+            writers[partition_key] = FormatFactory::instance().getOutputFormatParallelIfPossible(
+                configuration->format, *write_buffers[partition_key], sample_block, context, format_settings);
+        }
+
+        writers[partition_key]->write(getHeader().cloneWithColumns(part_chunk.getColumns()));
+    }
+}
+
+void IcebergStorageSink::onFinish()
+{
+    if (isCancelled())
+        return;
+
+    finalizeBuffers();
+    releaseBuffers();
+}
+
+void IcebergStorageSink::finalizeBuffers()
+{
+    for (const auto & [partition_key, _] : data_filenames)
+    {
+        try
+        {
+            writers[partition_key]->flush();
+            writers[partition_key]->finalize();
+        }
+        catch (...)
+        {
+            /// Stop ParallelFormattingOutputFormat correctly.
+            cancelBuffers();
+            releaseBuffers();
+            throw;
+        }
+
+        write_buffers[partition_key]->finalize();
+    }
+    initializeMetadata();
+}
+
+void IcebergStorageSink::releaseBuffers()
+{
+    for (const auto & [partition_key, _] : data_filenames)
+    {
+        writers[partition_key].reset();
+        write_buffers[partition_key].reset();
+    }
+}
+
+void IcebergStorageSink::cancelBuffers()
+{
+    for (const auto & [partition_key, _] : data_filenames)
+    {
+        if (writers[partition_key])
+            writers[partition_key]->cancel();
+        if (write_buffers[partition_key])
+            write_buffers[partition_key]->cancel();
+    }
+}
+
+void IcebergStorageSink::initializeMetadata()
+{
+    auto manifest_list_name = filename_generator.generateManifestListName();
+    auto metadata_name = filename_generator.generateMetadataName();
+
+    Int64 parent_snapshot = metadata->getValue<Int64>(Iceberg::f_current_snapshot_id);
+    auto new_snapshot = MetadataGenerator(metadata).generateNextMetadata(
+        manifest_list_name, parent_snapshot, 1, total_rows, total_chunks_size, static_cast<Int32>(data_filenames.size()));
```

Comment:
Why `added_files=1`? What if we written partitioned data and have multiple data files?